File: Kickstart19/C.java
import java.io.*;
import java.util.*;
class Abc
{
	static class FastReader 
	{ 
	    BufferedReader br; 
	    StringTokenizer st; 
	
	    public FastReader() 
	    { 
	        br = new BufferedReader(new
	                 InputStreamReader(System.in)); 
	    } 
	
	    String next() 
	    { 
	        while (st == null || !st.hasMoreElements()) 
	        { 
	            try
	            { 
	                st = new StringTokenizer(br.readLine()); 
	            } 
	            catch (IOException  e) 
	            { 
	                e.printStackTrace(); 
	            } 
	        } 
	        return st.nextToken(); 
	    } 
	
	    int nextInt() 
	    { 
	        return Integer.parseInt(next()); 
	    } 
	
	    long nextLong() 
	    { 
	        return Long.parseLong(next()); 
	    } 
	
	    double nextDouble() 
	    { 
	        return Double.parseDouble(next()); 
	    } 
	
	    String nextLine() 
	    { 
	        String str = ""; 
	        try
	        { 
	            str = br.readLine(); 
	        } 
	        catch (IOException e) 
	        { 
	            e.printStackTrace(); 
	        } 
	        return str; 
	    } 
	} 
	public static void main(String args[])
	{
		FastReader in=new FastReader();
		int t=in.nextInt();
		for(int g=0;g<t;g++)
		{
			long n,k,x1,y1,c,d,e1,e2,f;
			n=in.nextLong();
			k=in.nextLong();
			x1=in.nextLong();
			y1=in.nextLong();
			c=in.nextLong();
			d=in.nextLong();
			e1=in.nextLong();
			e2=in.nextLong();
			f=in.nextLong();

			long x[]=new long[(int)n];
			long y[]=new long[(int)n];
			long ar[]=new long[(int)n];

			x[0]=x1;
			y[0]=y1;
			ar[0]=(x1+y1)%f;

			for(int i=1;i<n;i++)
			{
				x[i]=(c*x[i-1]+d*y[i-1]+e1)%f;
				y[i]=(d*x[i-1]+c*y[i-1]+e2)%f;
				ar[i]=(x[i]+y[i])%f;
			}

			long sum=(ar[0]*((n*k)%mod))%mod;
			long sm=k;
			for(int i=2;i<=n;i++)
			{
				long num=(i*((pow(i,k)-1+mod)%mod));
				sm+=(num*modInverse((i-1+mod)%mod))%mod;
				sm%=mod;
				sum+=(ar[i-1]*(((n+1-i+mod)%mod*sm)%mod)+mod)%mod;
				sum%=mod;
			}
			System.out.println("Case #"+(g+1)+": "+sum);
		}
	}
	static long mod=1000000007;
	public static long pow(long a,long b)
	{
		long result=1;
		while(b>0)
		{
			if(b%2==1)
			{
				result=(result*a)%mod;
			}
			a=(a*a)%mod;
			b/=2;
		}
		return result%mod;
	}
	static long xgcd,ygcd;
	public static long modInverse(long a)
	{
		gcd(a,mod);
		return(xgcd%mod+mod)%mod;
	}
	public static void gcd(long a,long b)
	{
		if(b==0)
		{
			xgcd=1;
			ygcd=0;
		}
		else
		{
			gcd(b,a%b);
			long temp=xgcd;
			xgcd=ygcd;
			ygcd=temp-((a/b)*ygcd);

		}
	}
}

File: Kickstart19/C.cpp
#include<iostream>
using namespace std;
#define ll long long

long long mod=1000000007;

ll pow(ll a,ll b)
{
	ll result=1;
	while(b>0)
	{
		if(b%2==1)
		{
			result=(result*a)%mod;
		}
		a=(a*a)%mod;
		b/=2;
	}
	return result;
}

ll xgcd,ygcd;

ll gcd(ll a,ll b)
{
	if(b==0)
	{
		xgcd=1;
		ygcd=0;
	}
	else
	{
		gcd(b,a%b);
		ll temp=xgcd;
		xgcd=ygcd;
		ygcd=temp-((a/b)*ygcd);
	}
}

ll modInverse(ll a)
{
	gcd(a,mod);
	return (xgcd%mod+mod)%mod;
}

int main()
{
	int t;
	cin>>t;
	for(int g=0;g<t;g++)
	{
		ll n,k,x1,y1,c,d,e1,e2,f;
		cin>>n>>k>>x1>>y1>>c>>d>>e1>>e2>>f;
		ll x[n];
		ll y[n];
		ll ar[n];

		x[0]=x1;
		y[0]=y1;
		ar[0]=(x1+y1)%f;
		//cout<<ar[0]<<endl;

		for(int i=1;i<n;i++)
		{
			x[i]=((c*x[i-1])%f+(d*y[i-1])%f+e1+f)%f;
			y[i]=((d*x[i-1])%f+(c*y[i-1])%f+e2+f)%f;
			ar[i]=(x[i]+y[i]+2*f)%f;
			//cout<<ar[i]<<endl;
		}
		ll sum=(ar[0]*((k*n)%mod))%mod;
		ll sm=k;
		cout<<1<<" Sm = "<<sm<<endl;
		for(int i=2;i<=n;i++)
		{
			ll num=(i*((pow(i,k)-1+mod)%mod))%mod;// i(i^k-1)
			sm+=(num*modInverse((i-1+mod)%mod))%mod;//(i(i^k-1))/(i-1)
			sm%=mod;
			cout<<i<<" Sm = "<<sm<<endl;
			sum+=(ar[i-1]*((sm*(n+1-i))%mod))%mod;
			sum%=mod;
		}
		cout<<"Case #"<<g+1<<": "<<sum<<endl;
	}
	return 0;
}